Sources/NIOHPACK/HPACKHeader.swift

```diff
@@ -0,0 +1,421 @@
+    /// - Returns: A list of the values for that header field name.
+    public subscript(canonicalForm name: String) -> [String] {
+        let result = self[name]
+        guard result.count > 0 else {
+            return []
+        }
+        
+        // It's not safe to split Set-Cookie on comma.
+        guard name.lowercased() != "set-cookie" else {
+            return result
+        }
+        
+        return result.flatMap { $0.split(separator: ",") }.map { String($0.trimWhitespace()) }
+    }
+    
+    /// Special internal function for use by tests.
+    internal subscript(position: Int) -> (String, String) {
+        precondition(position < self.headers.endIndex, "Position \(position) is beyond bounds of \(self.headers.endIndex)")
+        let header = self.headers[position]
+        let name = self.string(idx: header.name)
+        let value = self.string(idx: header.value)
+        return (name, value)
+    }
+    
+    /// Enables encoders to directly enumerate the header views themselves without converting via `String`.
+    internal var headerIndices: [HPACKHeader] {
+        return self.headers
+    }
+    internal func view(of index: HPACKHeaderIndex) -> ByteBufferView {
+        return self.buffer.viewBytes(at: index.start, length: index.length)
+    }
+}
+
+extension HPACKHeaders: Sequence {
+    public typealias Element = (name: String, value: String, indexable: HPACKIndexing)
+    
+    /// An iterator of HTTP header fields.
+    ///
+    /// This iterator will return each value for a given header name separately. That
+    /// means that `name` is not guaranteed to be unique in a given block of headers.
+    public struct Iterator : IteratorProtocol {
+        private var headerParts: Array<(String, String, HPACKIndexing)>.Iterator
+        
+        init(headerParts: Array<(String, String, HPACKIndexing)>.Iterator) {
+            self.headerParts = headerParts
+        }
+        
+        public mutating func next() -> Element? {
+            return headerParts.next()
+        }
+    }
+    
+    public func makeIterator() -> HPACKHeaders.Iterator {
+        return Iterator(headerParts: headers.map { (self.string(idx: $0.name), self.string(idx: $0.value), $0.indexing) }.makeIterator())
+    }
+}
+
+extension HPACKHeaders: CustomStringConvertible {
+    public var description: String {
+        var headersArray: [(HPACKIndexing, String, String)] = []
+        headersArray.reserveCapacity(self.headers.count)
+        
+        for h in self.headers {
+            headersArray.append((h.indexing, self.string(idx: h.name), self.string(idx: h.value)))
+        }
+        return headersArray.description
+    }
+}
+
+extension HPACKHeaders: Equatable {
+    public static func == (lhs: HPACKHeaders, rhs: HPACKHeaders) -> Bool {
+        guard lhs.headers.count == rhs.headers.count else {
+            return false
+        }
+        
+        let lhsNames = Set(lhs.names.map { lhs.string(idx: $0).lowercased() })
```

Comment:
think we should use `.lazy.map` here as otherwise you'll first build (allocate & initialise) an `Array` and then build a `Set` throwing the newly built `Array` away again.